kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ZookeeperDiscoveryClient.scala

```diff
@@ -385,12 +381,26 @@ class ZookeeperDiscoveryClient(conf: KyuubiConf) extends DiscoveryClient {
     localServiceNode
   }
 
-  class DeRegisterWatcher(instance: String, serviceDiscovery: ServiceDiscovery) extends Watcher {
+  private def watchNode(): Unit = {
+    if (zkClient.checkExists
+        .usingWatcher(watcher.asInstanceOf[Watcher]).forPath(serviceNode.getActualPath) == null) {
+      // No node exists, throw exception
+      throw new KyuubiException(s"Unable to create znode for this Kyuubi " +
+        s"instance[${watcher.instance}] on ZooKeeper.")
+    }
+  }
+
+  class DeRegisterWatcher(val instance: String, serviceDiscovery: ServiceDiscovery)
+    extends Watcher {
     override def process(event: WatchedEvent): Unit = {
       if (event.getType == Watcher.Event.EventType.NodeDeleted) {
-        warn(s"This Kyuubi instance ${instance} is now de-registered from" +
-          s" ZooKeeper. The server will be shut down after the last client session completes.")
+        warn(s"This Kyuubi instance $instance is now de-registered from" +
+          " ZooKeeper. The server will be shut down after the last client session completes.")
+        ZookeeperDiscoveryClient.this.deregisterService()
         serviceDiscovery.stopGracefully()
+      } else if (event.getType == Watcher.Event.EventType.NodeDataChanged) {
```

Comment:
after we fix the re-create node issue during stopGracefully, when will NodeDataChanged happen ?

Reply:
Not sure in what scenario the `NodeDataChanged` event occurs.
I found that the znode mtime of some users in the production environment has changed, so I added logs and re-watched.
If we use `zkCli.sh` set znode data, a `NodeDataChanged` event will be sent,

Reply:
Yes.

The `DeRegisterWatcher` implemented by `HiveServer2` only handles the `NodeDeleted` event and does not handle `NodeDataChanged`.

https://github.com/apache/hive/blob/f940a0963f68b790e468f6e3f77721cc4edb2161/service/src/java/org/apache/hive/service/server/HiveServer2.java#L635

If the watcher implemented by `HDFS` has a `NodeDataChanged` event, it will re-watch.

https://github.com/apache/hadoop/blob/d78b0b39a6360c7131662c818dd498fc644484d7/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ha/ActiveStandbyElector.java#L688-L689

Reply:
looks reasonable